### ISSUE TITLE
Uncaught TypeError: $element.data(...).tip is not a function

### ISSUE THREAD

**Issue by dsgn1graphics** [sorich87/bootstrap-tour#619]
I'm not sure what changed but all of a sudden I'm seeing this error in the console output. I'm using the current version available at at https://cdnjs.cloudflare.com/ajax/libs/bootstrap-tour/0.11.0/js/bootstrap-tour.js

```js
Uncaught TypeError: $element.data(...).tip is not a function
    at Tour._showPopover (bootstrap-tour.js?v=:586)
    at showPopoverAndOverlay (bootstrap-tour.js?v=:352)
    at HTMLBodyElement.<anonymous> (bootstrap-tour.js?v=:712)
    at HTMLBodyElement.d.complete (jquery.min.js:3)
    at i (jquery.min.js:2)
    at Object.fireWith [as resolveWith] (jquery.min.js:2)
    at i (jquery.min.js:3)
    at Function.r.fx.tick (jquery.min.js:3)
    at eb (jquery.min.js:3)
```

Here my 2min debug from the js console. `tip` returns an html node element, and it is not initialized as a jQuery element which means the following lines of code throw exceptions after variable assignment

```js
$element.data('bs.popover').tip.toString()
=> "[object HTMLDivElement]" 
```

Here's my temp fix, it would be nice to get this in the next release, so that I can switch back to using the CDN url instead of my locally modified copy.

line: 586
```js
$tip = $element.data('bs.popover') ? $element.data('bs.popover').tip() : $element.data('popover').tip();
```

changed to 

```js
$tip = $element.data('bs.popover') ? $($element.data('bs.popover').tip): $($element.data('popover').tip);
```

**Comment by sajumani** [comment]
am also getting same error..

bootstrap-tour.js:597 Uncaught TypeError: $element.data(...).tip is not a function
    at Tour._showPopover (bootstrap-tour.js:597)
    at Tour._showPopoverAndOverlay (bootstrap-tour.js:552)
    at Tour._showPopoverAndOverlay (bootstrap-tour.js:22)
    at HTMLBodyElement.<anonymous> (bootstrap-tour.js:717)
    at HTMLBodyElement.r.complete (jquery-1.9.1.min.js:5)
    at c (jquery-1.9.1.min.js:3)
    at Object.fireWith [as resolveWith] (jquery-1.9.1.min.js:3)
    at u (jquery-1.9.1.min.js:5)
    at b.fx.tick (jquery-1.9.1.min.js:5)



is it work with bootstrap v4 ??

**Comment by azizmashkour** [comment]
I think bootstrap v4 uses getTipElement() instead of tip(), so use $element.data('bs.popover').getTipElement() and it will work.

**Comment by sajumani** [comment]
Thanks, 

it fixed but am getting another error

Uncaught TypeError: Cannot read property 'offsetWidth' of undefined

i think, they needs to rewrite it for Bootstrap v4

**Comment by azizmashkour** [comment]
Are you using the master's version of bootstraptour? Anyway, the compatibility with bootstrap v4 is in progress and will available soon.

**Comment by sajumani** [comment]
yes am using master version....

ok, please update ASAP, more people waiting for this...

**Comment by nishantpant** [comment]
Yes, please let us know when v4 compatible version is released. Looking forward to it.

**Comment by mysyfy** [comment]
+1 waiting BS4 compatibility upgrade.

**Comment by sorich87** [comment]
If you're not using the `orphan` setting, latest version from master should be compatible.

**Comment by sajumani** [comment]
@sorich87 am using latest version, which is v12.0 but the title and description not showing..

and the demo page also not updated with latest version... it would be good if updated all correct info

**Comment by sorich87** [comment]
I meant latest master, not latest released version

**Comment by sajumani** [comment]
@sorich87 no error but popover title and content not displaying...

**Comment by sajumani** [comment]
@sorich87 i fixed the issue... you template not compatible with v4... this should be 

template: '<div class="popover" role="tooltip"> <div class="arrow"></div> <h3 class="popover-title"></h3> <div class="popover-content"></div> <div class="popover-navigation"> <div class="btn-group"> <button class="btn btn-sm btn-secondary" data-role="prev">&laquo; Prev</button> <button class="btn btn-sm btn-secondary" data-role="next">Next &raquo;</button> <button class="btn btn-sm btn-secondary" data-role="pause-resume" data-pause-text="Pause" data-resume-text="Resume">Pause</button> </div> <button class="btn btn-sm btn-secondary" data-role="end">End tour</button> </div> </div>',

**Comment by dimerman** [comment]
any idea how can I use the latest master with yarn ?

**Comment by jkyadav** [comment]
Is the BS4 compatible version released?

**Comment by JYovan** [comment]
Is this compatible with Bootstrap v4.0.0 ? 06/Jun/2018

**Comment by iamujala** [comment]
bootstrap-tour.min.js:22 Uncaught TypeError: t(...).stop is not a function
    at o._scrollIntoView (bootstrap-tour.min.js:22)
    at o.<anonymous> (bootstrap-tour.min.js:22)
    at o._callOnPromiseDone (bootstrap-tour.min.js:22)
    at o.showStep (bootstrap-tour.min.js:22)
    at o.init (bootstrap-tour.min.js:22)
    at index.html:111
o._scrollIntoView @ bootstrap-tour.min.js:22
(anonymous) @ bootstrap-tour.min.js:22
o._callOnPromiseDone @ bootstrap-tour.min.js:22
o.showStep @ bootstrap-tour.min.js:22
o.init @ bootstrap-tour.min.js:22
(anonymous) @ index.html:111

Facing this issue. Please help.

**Comment by carlosalvidrez** [comment]
Did this ever worked with Bootstrap 4?

**Comment by OmarMakled** [comment]
You can use `bootstrap-tour-standalone` 

The order is important 

```
window.$ = window.jQuery = require('jquery')
require('bootstrap') // Bootstrap 4 
window.Tour = require("../node_modules/bootstrap-tour/build/js/bootstrap-tour-standalone");
```

if you import `bootstrap` after `bootstrap-tour-standalone` you will get `Uncaught TypeError: $element.data(...).tip is not a function`

**Comment by henry9162** [comment]
> I meant latest master, not latest released version

Thanks a lot for this, 3 years later in 2020, it still as solving problems. Gracias!

**Comment by vijayan108** [comment]
> I'm closing this issue, no error. I dug into this a little bit more and used the demo that's running on bootstraptour.com to figure out what was wrong.
> 
> I didn't realize that localStorage is being used to maintain the tour state. I noticed the option in the api docs but figured that it was just an option, and wasn't actually initialized by default. I'll be turning that off and storing states on the backend with a server session key and metadata attached to my users.

can u tell what option u have enabled because im facing this error
